### DISCUSSION TITLE
I'd need some help with `run-shell-command`

### DISCUSSION THREAD

**Discussion by dariooddenino** [helix-editor/helix#3132] [Q&A]
Hi!

I'd like to pass commands from helix to a zellij session, but I'm encountering a couple of roadblocks:

### 1 `:sh zellij -s ONE action 'NewTab:'` 
This command should create a new tab on the session named "ONE". 
The command fails, and the log shows:
```
[ERROR] Shell error: failed to open terminal:
ENOTTY: Not a typewriter
```

I'm not familiar with how terminals work, but I'm guessing helix is not getting the return code/output it's expecting. In emacs I can run async shell commands ignoring completely the output and return code, and it would be nice to have it here too.

### 2 `:sh zellij -s ONE action 'WriteChars: "a"'`
This command should write "a" on the session named "ONE.
The command fails with this log message:
```
[ERROR] Shell error: error: Found argument 'a' which wasn't expected, or isn't valid in this context
```
I get the same error even if I omit the double quotes around `a` (which isn't valid syntax for that command).

I'm confused by this one, because it appears that helix handles quotes correctly.
For example `:sh echo 'WriteChars: "a"' > foo.txt` works fine.

Thanks!

**Comment by dariooddenino** [comment]
I had some time to play with this issue.

### 1. The first problem is causes by this:
https://github.com/helix-editor/helix/blob/61856f1d64557a053f9160698ab8751bf01d76ab/helix-term/src/commands.rs#L4544-L4550

I've made a quick test calling `.stdin()` on `Command` only when the `input` arg to `shell_impl` is `Some`, and everything seems to work fine.
I wonder if this could be applied as a temporary solution.

### 2. I can avoid the second problem by escaping the double quotes

So `:sh zellij -s ONE action \"WriteChars: \"a\"\"`. 
 
I guess it could be solved in a more elegant way but it's good enough for now.

**Comment by AceofSpades5757** [comment]
1. The not a typewriter has to do with user input. The command you wrote may have required user input that is not yet implemented by Helix.
2. This looks like something to do with that specific command. I ran through what I could and didn't find any issues with quoting. As in, anything I input acted just like if I had put it in a normal terminal.

Note: There's currently a pull request to make impl_shell concurrent which may solve applicable async issues: https://github.com/helix-editor/helix/pull/3180

**Reply by AceofSpades5757** [reply]
Specifically, when looking online, it's `scanf` (or something similar) used in C libraries. When attempting to replicate the issue with similar tools, I got various errors.

**Reply by dariooddenino** [reply]
The command doesn't require any user input, but it could also be that there's something wrong on its implementation.

I think all quotes (unless escaped) are removed here:
https://github.com/helix-editor/helix/blob/master/helix-term/src/commands/typed.rs#L2076

And after that the string is split at each space.
This means, for this command, that this section `'WriteChars: "a"'` is sent as two separate arguments (`WriteChars:` and `a`), making the command fail.
